### PR TITLE
Pointwise, Pairwise and Listwise Losses

### LINKED ISSUE #139: Losses - Mechanism to compute the avg pairwise losses for a list with one positive and many negative samples

### LINKED ISSUE #140: Pairwise Losses - BPR, BPR-max, TOP1, TOP1-max Log loss, hinge loss, adaptive hinge loss

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
This PR provides the implementation of a comprehensive list of pairwise losses used in RecSys:
- [x] BPR loss
- [x] BPR-max loss
- [x] TOP1 loss
- [x] TOP1-max loss
- [x] Log loss
- [x] Hinge loss
- [x] Adaptive Hinge loss

We also wrap these pointwise and listwise losses from TF, so that they can be used as a string in our loss registry
- [x] Mean Squared error (pointwise)
- [x] Binary cross entropy (pointwise)
- [x] Categorical cross entropy (listwise)
- [x] Sparse categorical cross entropy (listwise)

Closes #139, closes #140

### CODE REVIEW

File: merlin_models/tf/__init__.py [hunk 1]
```diff
@@ -85,6 +85,16 @@
     NextItemPredictionTask,
     YoutubeDNNRetrieval,
 )
+from .prediction.losses import (
+    AdaptiveHinge_Loss,
```
Comment (marcromeyn) [inline merlin_models/tf/__init__.py:89]: Our public API might get too cluttered when we put all of this here. Maybe we should have the import be: `merlin_models.loss.HingLoss`. Also I think we should remove the `_` since in python it’s camel-case for classes and snake-case for functions & variables.

File: merlin_models/tf/prediction/losses.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+# limitations under the License.
+#
+
+from typing import Tuple
+
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+
+
+class PairwiseLoss(Loss):
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:23]: Maybe we should have a `loss` package, and then have the pair_wise.py that has all the pair-wise losses?

File: merlin_models/tf/prediction/losses.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+            Loss per example
+        """
+        positives_scores, negatives_scores = self._get_positives_negatives_scores(y_true, y_pred)
+        sub = negatives_scores - positives_scores
+        loss = tf.nn.sigmoid(sub) + tf.nn.sigmoid(tf.square(negatives_scores))
+        return loss
+
+
+class TOP1v2_Loss(PairwiseLoss):
+    """An adapted version of the TOP pairwise loss proposed in [1]_, but following the the
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:179]: Typo: `the the`

File: merlin_models/tf/prediction/losses.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Tuple
+
+import tensorflow as tf
+from tensorflow.keras.losses import Loss
+
+
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:22]: Should we create a loss registry so people can use these using a string?

File: merlin_models/tf/prediction/losses.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    """The TOP1-max pairwise loss proposed in [1]_
+
+    References
+    ----------
+    .. [1] Hidasi, Balázs, and Alexandros Karatzoglou. "Recurrent neural networks with top-k gains
+       for session-based recommendations." Proceedings of the 27th ACM international conference on
+       information and knowledge management. 2018. https://arxiv.org/abs/1706.03847
+    """
+
+    def call(self, y_true: tf.Tensor, y_pred: tf.Tensor):
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:228]: Needs output-type.

File: merlin_models/tf/prediction/losses.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    current GRU4Rec implementation [2]_.
+
+    References:
+    .. [1] B. Hidasi, A. Karatzoglou, L. Baltrunas, and D. Tikk, “Session-based recommendations
+        with recurrent neural networks,” in Proceedings of Fourth International Conference on
+        Learning Representations (ICLR’16), 2016.
+    .. [2] https://arxiv.org/abs/1511.06939
+    """
+
+    def call(self, y_true: tf.Tensor, y_pred: tf.Tensor):
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:189]: Needs output-type.

File: merlin_models/tf/prediction/losses.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    """The TOP pairwise loss proposed in [1]_
+
+    References
+    ----------
+    .. [1] B. Hidasi, A. Karatzoglou, L. Baltrunas, and D. Tikk, “Session-based recommendations
+       with recurrent neural networks,” in Proceedings of Fourth International Conference on
+       Learning Representations (ICLR’16), 2016. https://arxiv.org/abs/1511.06939
+    """
+
+    def call(self, y_true: tf.Tensor, y_pred: tf.Tensor):
```
Comment (marcromeyn) [inline merlin_models/tf/prediction/losses.py:157]: Needs output-type.

File: merlin_models/tf/__init__.py (hunk 1 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/__init__.py]: Done

File: merlin_models/tf/prediction/losses.py (hunk 2 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Done

File: merlin_models/tf/prediction/losses.py (hunk 3 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Fixed

File: merlin_models/tf/prediction/losses.py (hunk 4 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Done

File: merlin_models/tf/prediction/losses.py (hunk 5 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Done

File: merlin_models/tf/prediction/losses.py (hunk 6 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Fixed

File: merlin_models/tf/prediction/losses.py (hunk 7 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/losses.py]: Fixed

File: merlin_models/tf/losses/loss_base.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+from merlin_standard_lib import Registry, RegistryMixin
+
+LossType = Union[str, tf.keras.losses.Loss]
+
+loss_registry: Registry = Registry.class_registry("tf.losses")
+
+# Registering keras pointwise losses
+loss_registry.register("mse")(tf.keras.losses.MeanSquaredError)
+loss_registry.register("binary_crossentropy")(tf.keras.losses.BinaryCrossentropy)
```
Comment (marcromeyn) [inline merlin_models/tf/losses/loss_base.py:29]:
```suggestion
loss_registry.registry.register_with_multiple_names(“binary_crossentropy”, “bce")(tf.keras.losses.BinaryCrossentropy)
```

File: merlin_models/tf/losses/loss_base.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import tensorflow as tf
+
+from merlin_standard_lib import Registry, RegistryMixin
+
+LossType = Union[str, tf.keras.losses.Loss]
+
+loss_registry: Registry = Registry.class_registry("tf.losses")
+
+# Registering keras pointwise losses
+loss_registry.register("mse")(tf.keras.losses.MeanSquaredError)
```
Comment (marcromeyn) [inline merlin_models/tf/losses/loss_base.py:28]:
```suggestion
loss_registry.registry.register_with_multiple_names(“mean_squared_error”, "mse")(tf.keras.losses.MeanSquaredError)
```

File: merlin_models/tf/losses/listwise_losses.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    manually in the output layer). It also adds support to
+    the loss_registry, so that the loss can be defined by the user
+    by a string alias name.
+    """
+
+    def __init__(self, from_logits=True, **kwargs):
+        super().__init__(from_logits=from_logits, **kwargs)
+
+
+@LossRegistryMixin.registry.register("categ_crossentropy")
```
Comment (marcromeyn) [inline merlin_models/tf/losses/listwise_losses.py:38]: Let’s keep the names in-line with Keras here ([source](https://keras.io/api/losses/)).

```suggestion
@LossRegistryMixin.registry.register("categorical_crossentropy")
```

File: merlin_models/tf/losses/listwise_losses.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+# limitations under the License.
+#
+
+import tensorflow as tf
+from tensorflow.keras.losses import CategoricalCrossentropy, SparseCategoricalCrossentropy
+
+from .loss_base import LossRegistryMixin
+
+
+@LossRegistryMixin.registry.register("sparse_categ_crossentropy")
```
Comment (marcromeyn) [inline merlin_models/tf/losses/listwise_losses.py:23]: Same as comment below

File: merlin_models/tf/losses/loss_base.py (hunk 8 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/losses/loss_base.py]: Done

File: merlin_models/tf/losses/loss_base.py (hunk 9 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/losses/loss_base.py]: Done

File: merlin_models/tf/losses/listwise_losses.py (hunk 10 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/losses/listwise_losses.py]: Done

File: merlin_models/tf/losses/listwise_losses.py (hunk 11 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/losses/listwise_losses.py]: Done
